src/vs/workbench/services/editor/common/customEditorLabelService.ts

```diff
@@ -0,0 +1,185 @@
+	readonly onDidChange = this._onDidChange.event;
+
+	private patterns: ICustomEditorLabelPattern[] = [];
+	private enabled = true;
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IWorkspaceContextService private readonly workspaceContextService: IWorkspaceContextService,
+	) {
+		super();
+
+		this.storeEnablementState();
+		this.storeCustomPatterns();
+
+		this.registerListernes();
+	}
+
+	private registerListernes(): void {
+		this._register(this.configurationService.onDidChangeConfiguration(e => {
+			// Cache the enabled state
+			if (e.affectsConfiguration(CustomEditorLabelService.SETTING_ID_ENABLED)) {
+				const oldEnablement = this.enabled;
+				this.storeEnablementState();
+				if (oldEnablement !== this.enabled && this.patterns.length > 0) {
+					this._onDidChange.fire();
+				}
+			}
+
+			// Cache the patterns
+			else if (e.affectsConfiguration(CustomEditorLabelService.SETTING_ID_PATTERNS)) {
+				this.storeCustomPatterns();
+				this._onDidChange.fire();
+			}
+		}));
+	}
+
+	private storeEnablementState(): void {
+		this.enabled = this.configurationService.getValue<boolean>(CustomEditorLabelService.SETTING_ID_ENABLED);
+	}
+
+	private storeCustomPatterns(): void {
+		this.patterns = [];
+		const customLabelPatterns = this.configurationService.getValue<ICustomEditorLabelObject>(CustomEditorLabelService.SETTING_ID_PATTERNS);
+		for (const pattern in customLabelPatterns) {
+			const template = customLabelPatterns[pattern];
+
+			const isAbsolutePath = isAbsolute(pattern);
+			const parsedPattern = parseGlob(pattern);
+
+			this.patterns.push({ pattern, template, isAbsolutePath, parsedPattern });
+		}
+
+		this.patterns.sort((a, b) => this.patternWeight(b.pattern) - this.patternWeight(a.pattern));
+	}
+
+	private patternWeight(pattern: string): number {
+		let weight = 0;
+		for (const fragment of pattern.split('/')) {
+			if (fragment === '**') {
+				weight += 1;
+			} else if (fragment === '*') {
+				weight += 10;
+			} else if (fragment.includes('*') || fragment.includes('?')) {
+				weight += 50;
+			} else if (fragment !== '') {
+				weight += 100;
+			}
+		}
+
+		return weight;
+	}
+
+	getName(resource: URI): string | undefined {
+		if (!this.enabled) {
+			return undefined;
+		}
+		return this.applyPatterns(resource);
+	}
+
+	private applyPatterns(resource: URI): string | undefined {
+		if (this.patterns.length === 0) {
+			return undefined;
+		}
+
+		const root = this.workspaceContextService.getWorkspaceFolder(resource);
+		for (const pattern of this.patterns) {
+			if (this.matchPattern(root, resource, pattern)) {
+				return this.applyTempate(pattern.template, resource);
+			}
+		}
+
+		return undefined;
+	}
+
+	private matchPattern(root: IWorkspaceFolder | null, resource: URI, pattern: ICustomEditorLabelPattern): boolean {
+		const relevantPath = !root || pattern.isAbsolutePath ? resource.fsPath : relativePath(root.uri, resource) ?? resource.path;
```

Comment:
You keep computing the relative path per pattern, but I would move this out to the `applyPatterns` and only do this if its also needed by the pattern.

Reply:
Also, why do you jump between `resource.fsPath` and `resource.path`?